lasagne/objectives.py

```diff
@@ -12,74 +14,136 @@ def mse(x, t):
     :returns:
         - output : the mean square error across all dimensions
     """
-    return T.mean((x - t) ** 2)
+    return (x - t) ** 2
 
 
-def crossentropy(x, t):
-    """Calculates the binary crossentropy mean across all dimentions,
-    i.e.  feature dimension AND minibatch dimension.
+class Objective(object):
+    _valid_aggregation = {None, 'mean', 'sum'}
 
-    :parameters:
-        - x : predicted values
-        - t : target values
+    """
+    Training objective
 
-    :returns:
-        - output : the mean binary cross entropy across all dimensions
+    The  `get_loss` method returns an expression that can be used for
+    training with a gradient descent approach.
     """
-    return T.mean(T.nnet.binary_crossentropy(x, t))
+    def __init__(self, input_layer, loss_function=mse, aggregation='mean'):
+        """
+        Constructor
 
+        :parameters:
+            - input_layer : a `Layer` whose output is the networks prediction
+                given its input
+            - loss_function : a loss function of the form `f(x, t)` that
+                returns a scalar loss given tensors that represent the
+                predicted and true values as arguments..
+            - aggregation : either:
+                - `'mean'` or `None` : the mean of the the elements of the
+                loss will be returned
+                - `'sum'` : the sum of the the elements of the loss will be
+                returned
+        """
+        self.input_layer = input_layer
+        self.loss_function = loss_function
+        self.target_var = T.matrix("target")
+        if aggregation not in self._valid_aggregation:
+            raise ValueError('aggregation must be \'mean\', \'sum\', '
+                             'or None, not {0}'.format(aggregation))
+        self.aggregation = aggregation
 
-def multinomial_nll(x, t):
-    """Calculates the mean multinomial negative-log-loss
+    def get_loss(self, input=None, target=None, aggregation=None,
+                 *args, **kwargs):
+        """
+        Get loss scalar expression
 
-    :parameters:
-        - x : (predicted) class probabilities; a theano expression resulting
-            in a 2D array; samples run along axis 0, class probabilities along
-            axis 1
-        - t : (correct) class probabilities; a theano expression resulting in
-            a 2D tensor that gives the class probabilities in its rows, OR
-            a 1D integer array that gives the class index of each sample
-            (the position of the 1 in the row in a 1-of-N encoding, or
-            1-hot encoding),
+        :parameters:
+            - input : (default `None`) an expression that results in the
+                input data that is passed to the network
+            - target : (default `None`) an expression that results in the
+                desired output that the network is being trained to generate
+                given the input
```

Comment:
The `aggregation` parameter needs to be documented (can be copied from the `MaskedObjective` below)